File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/data/network/NetworkService.kt
package com.runanywhere.sdk.data.network

import com.runanywhere.sdk.data.network.models.APIEndpoint

/**
 * Network service interface - equivalent to iOS NetworkService protocol
 * Enhanced with generic POST/GET methods and proper authentication support
 */
interface NetworkService {
    /**
     * POST request with JSON payload and typed response
     * Equivalent to iOS: func post<T: Encodable, R: Decodable>(_ endpoint: APIEndpoint, _ payload: T, requiresAuth: Bool) async throws -> R
     */
    suspend fun <T : Any, R : Any> post(
        endpoint: APIEndpoint,
        payload: T,
        requiresAuth: Boolean = true,
    ): R

    /**
     * GET request with typed response
     * Equivalent to iOS: func get<R: Decodable>(_ endpoint: APIEndpoint, requiresAuth: Bool) async throws -> R
     */
    suspend fun <R : Any> get(
        endpoint: APIEndpoint,
        requiresAuth: Boolean = true,
    ): R

    /**
     * POST request with raw data payload
     * Equivalent to iOS: func postRaw(_ endpoint: APIEndpoint, _ payload: Data, requiresAuth: Bool) async throws -> Data
     */
    suspend fun postRaw(
        endpoint: APIEndpoint,
        payload: ByteArray,
        requiresAuth: Boolean = true,
    ): ByteArray

    /**
     * GET request with raw data response
     * Equivalent to iOS: func getRaw(_ endpoint: APIEndpoint, requiresAuth: Bool) async throws -> Data
     */
    suspend fun getRaw(
        endpoint: APIEndpoint,
        requiresAuth: Boolean = true,
    ): ByteArray
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/data/network/network_service.dart
import 'dart:async';
import 'dart:typed_data';

import 'package:runanywhere/data/network/api_endpoint.dart';

/// Protocol defining the network service interface.
///
/// Matches iOS `NetworkService` protocol.
/// Allows for environment-based implementations (real vs mock).
abstract class NetworkService {
  /// Perform a POST request with typed payload and response.
  Future<T> post<T>(
    APIEndpoint endpoint,
    Object payload, {
    required bool requiresAuth,
    required T Function(Map<String, dynamic>) fromJson,
  });

  /// Perform a GET request with typed response.
  Future<T> get<T>(
    APIEndpoint endpoint, {
    required bool requiresAuth,
    required T Function(Map<String, dynamic>) fromJson,
  });

  /// Perform a raw POST request (returns raw bytes).
  Future<Uint8List> postRaw(
    APIEndpoint endpoint,
    Uint8List payload, {
    required bool requiresAuth,
  });

  /// Perform a raw GET request (returns raw bytes).
  Future<Uint8List> getRaw(
    APIEndpoint endpoint, {
    required bool requiresAuth,
  });

  /// Perform a POST with custom path (for parameterized endpoints).
  Future<T> postWithPath<T>(
    String path,
    Object payload, {
    required bool requiresAuth,
    required T Function(Map<String, dynamic>) fromJson,
  });

  /// Perform a GET with custom path (for parameterized endpoints).
  Future<T> getWithPath<T>(
    String path, {
    required bool requiresAuth,
    required T Function(Map<String, dynamic>) fromJson,
  });
}
